hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileReaderImpl.java

```diff
@@ -0,0 +1,284 @@
+        if (comparedNextBlockFirstKey >= 0) {
+          // Searches the block that may contain the lookup key based the starting keys of
+          // all blocks (sorted in the TreeMap of block index entries), using binary search.
+          // The result contains the greatest key less than or equal to the given key,
+          // or null if there is no such key.
+
+          Map.Entry<Key, BlockIndexEntry> floorEntry = dataBlockIndexEntryMap.floorEntry(key);
+          if (floorEntry == null) {
+            // Key smaller than the start key of the first block
+            return SEEK_TO_BACKWARDS;
+          }
+          currentDataBlockEntry = Option.of(floorEntry.getValue());
+          currentDataBlock = Option.empty();
+          currentPos.setOffset(
+              (int) currentDataBlockEntry.get().getOffset() + HFILEBLOCK_HEADER_SIZE);
+        }
+      }
+      if (!currentDataBlockEntry.get().getNextBlockKey().isPresent()) {
+        // This is the last data block.  Check against the last key.
+        if (fileInfo.getLastKey().isPresent()) {
+          int comparedLastKey = key.compareTo(fileInfo.getLastKey().get());
+          if (comparedLastKey > 0) {
+            currentDataBlockEntry = Option.empty();
+            currentDataBlock = Option.empty();
+            currentPos.setEof();
+            return SEEK_TO_EOF;
+          }
+        }
+      }
+
+      if (!currentDataBlock.isPresent()) {
+        currentDataBlock = Option.of(instantiateHFileDataBlock(currentDataBlockEntry.get()));
+      }
+
+      return currentDataBlock.get()
+          .seekTo(currentPos, key, (int) currentDataBlockEntry.get().getOffset());
+    }
+    if (compareCurrent == 0) {
+      return SEEK_TO_FOUND;
+    }
+    // Backward seek not supported
+    return SEEK_TO_BACKWARDS;
+  }
+
+  @Override
+  public boolean seekTo() throws IOException {
+    initializeMetadata();
+    if (trailer.getNumKeyValueEntries() == 0) {
+      currentPos.setEof();
+      return false;
+    }
+    // Move the current position to the beginning of the first data block
+    currentPos.setOffset(dataBlockIndexEntryMap.firstKey().getOffset() + HFILEBLOCK_HEADER_SIZE);
+    currentPos.unsetEof();
+    currentDataBlockEntry = Option.of(dataBlockIndexEntryMap.firstEntry().getValue());
+    // The data block will be read when {@link #getKeyValue} is called
+    currentDataBlock = Option.empty();
+    return true;
+  }
+
+  @Override
+  public boolean next() throws IOException {
+    if (currentPos.isValid()) {
+      if (!currentDataBlock.isPresent()) {
+        currentDataBlock = Option.of(instantiateHFileDataBlock(currentDataBlockEntry.get()));
+      }
+      if (currentDataBlock.get().next(currentPos, (int) currentDataBlockEntry.get().getOffset())) {
+        // The position is advanced by the data block instance
+        return true;
+      }
+      currentDataBlockEntry = getNextBlockIndexEntry(currentDataBlockEntry.get());
+      currentDataBlock = Option.empty();
```

Comment:
should `currentDataBlock` be updated here when skipping tot he next block index entry? or L216-218 below handles it?

Reply:
L216-218 handles that.  The loading of a data block is deferred as much as possible until the key value pair needs to be read.